packages/w3up-client/src/client.js

```
@@ -311,4 +311,81 @@ export class Client extends Base {
+    if (!options.shards) {
+      // Remove association of content CID with selected space.
+      try {
+        await this.capability.upload.remove(contentCID)
+        /* c8 ignore start */
+      } catch (/** @type {any} */ err) {
+        throw new Error(
+          `remove failed for ${contentCID}: ${err.message ?? err}`
+        )
+      }
+      return
+    }
+
+    // Get shards associated with upload.
+    let upload
+    try {
+      upload = await this.capability.upload.get(contentCID)
+      /* c8 ignore start */
+    } catch (/** @type {any} */ err) {
+      throw new Error(`remove failed for ${contentCID}: ${err.message ?? err}`)
+    }
+    /* c8 ignore stop */
+
+    // Check if we have information about the shards to remove
+    if (!upload.root) {
+      throw new Error(
+        `⁂ upload not found. could not determine shards to remove.`
+      )
+    }
+    /* c8 ignore start */
+    if (!upload.shards || !upload.shards.length) {
+      return
+    }
+    /* c8 ignore stop */
+
+    // Remove shards
+    await Promise.allSettled(
+      upload.shards.map(async (shard) => {
+        try {
+          await this.capability.store.remove(shard)
+        } catch (/** @type {any} */ error) {
+          /* c8 ignore start */
+          // If not found, we can tolerate error as it may be a consecutive call for deletion where first failed
+          if (error?.cause?.name !== 'StoreItemNotFound') {
+            throw new Error(
+              `remove failed for shard ${shard.link()} from ${contentCID}: ${
+                error?.message
+              }`
+            )
+          }
+          /* c8 ignore stop */
+        }
+      })
+    )
+
+    // Remove association of content CID with selected space.
+    try {
+      await this.capability.upload.remove(contentCID)
+      /* c8 ignore start */
+    } catch (/** @type {any} */ err) {
+      throw new Error(`remove failed for ${contentCID}: ${err.message ?? err}`)
```

Again, I would just let this throw - we don't have anything to add here that the caller does not already know.